### PR TITLE
No more Jquery and semantic.JS!

### PR DESCRIPTION
**Pull request by samelhusseini**
Quite a few things going on here so let me explain. 

The only reason we're using Jquery in the editor is because Semantic.JS requires it but as it turns out many of our components are using React components and we've moved away from Jquery with the exception of: 
- Modal
- Dropdown
- Dimmer
- Loader

Using Jquery with React is an antipattern also and by removing the need for semantic.js we're able to get rid of Jquery as well which should speed  things up quite a bit. 
This change does just that. Now by removing semantic we also lose some of the work we did to make the semantic dimmer and modal accessible. It turns out however that work wasn't entirely complete and to spec, and there were a few holes in it, so I've gone ahead and re-implemented the Semantic Dropdown in react and made it accessible and to spec. 
The modal however (we were using React already), but the accessibility stuff was an addon and did direct DOM manipulation to achieve that. I've gone ahead instead and pulled in a small library that's fairly widely used called https://github.com/reactjs/react-modal. It's accessible out of the box, and I've adapted the semantic CSS rendering to that. 
For the dimmer and Loader I've implemented custom React components for those too. 

Now to the other part of the story. We use the helpers in core all over the place for three things: notifications, loader, and prompt dialogs (confirm and prompt). I didn't want to make changes all over the place so I've kept those. I've created a ConfirmDialog that utilizes the React Modal (instead of Jquery) but uses the same API to call a confirm. Since it's Async I've put it in its own wrapper and it's managed in its own node, ie: it renders itself when first needed, and maintains that dom element for subsequent dialogs. (Perhaps we can refactor this a little more down the track)
Same thing with notifications and loaders, kept the calling API but the internals have changed to use their own react doms. 

Finally, I've just gotten rid of Jquery dom manipulations all over the place, most of it was pretty straightforward. The simulator is still directly managing it's own DOM (frames). Another one that could be refactored a little more down the track. I've implemented the simulator entry / exit transition by hand (it's still authored in CSS, just executed without a library)

I've also updated React to the latest version and a bunch of other libraries as part of this change.
Change still needs to be tested cross browser, etc.

Test areas: 
- Accessibility of Modal (keyboard navigation)
- Accessibility of Dropdown (keyboard navigation)
- Key components: Home, Share, Add package, Blocks/JS, Sidedocs, Project settings, Full screen
- Less used areas: Extensions, Tracing, Serial editor, tutorial mode, youtube videos?
- Test all targets (looking at any use of core.requestAsync or use of Jquery in editor.js
- Test different views: mobile, tablet, computer
- **Ensure loadings correctly prevent using the editor in the background via keyboard navigation**

Tested: 
- [x] Adafruit extensions

Known issues: 
- [x] Home screen accessibility
- [x] Modal for package extensions

GG

### CODE REVIEW

**Comment by samelhusseini** [comment]
Note: I also investigating pulling in https://github.com/Semantic-Org/Semantic-UI-React as part of this change. And I got pretty far using it for most purposes in our editor but it's accessibility story was lacking, and it was difficult to adapt. 
I also ran into issues treeshaking the components that weren't being used out, so it bloated our binary.

File: package.json [hunk 1]
```diff
@@ -76,14 +76,14 @@
     "highlight.js": "^9.7.0",
     "lzma": "^2.3.2",
     "marked": "^0.3.9",
-    "fuse.js": "2.6.1",
+    "fuse.js": "^3.2.0",
```
Comment (pelikhan) [inline package.json]: Any api change?

File: webapp/src/app.tsx [hunk 2]
```diff
@@ -921,41 +841,47 @@ export class ProjectView
...
-        // clear the hash
-        pxt.BrowserUtils.changeHash("", true);
-        this.setState({ home: true });
-        this.allEditors.forEach(e => e.setVisible(false));
-        this.homeLoaded();
-    }
-
-    private homeLoaded() {
-        pxt.tickEvent('app.home');
+    importFile(file: File) {
```
Comment (pelikhan) [inline webapp/src/app.tsx]: Maybe this should move too to importhelpers

**Comment by samelhusseini** [comment]
Also note, we dont have a react app in our docs so I didnt refactor it out from pxtrunner or our docs, so semantic and jquery are still in our project just not the editor

File: webapp/src/codecard.tsx [hunk 3]
```diff
@@ -16,11 +17,7 @@ export class CodeCardView extends React.Component<pxt.CodeCard, CodeCardState> {
         this.state = {};
     }
 
-    componentDidUpdate() {
-        ($('.ui.embed') as any).embed();
```
Comment (pelikhan) [inline webapp/src/codecard.tsx]: Do we support YouTube video embeds?

File: webapp/src/simulator.ts [hunk 4]
```diff
@@ -221,23 +225,23 @@ export function run(pkg: pxt.MainPackage, debug: boolean, res: pxtc.CompileResul
 
 export function mute(mute: boolean) {
     driver.mute(mute);
-    $debugger.empty();
+    while (debuggerDOM.firstChild) debuggerDOM.removeChild(debuggerDOM.firstChild);
```
Comment (pelikhan) [inline webapp/src/simulator.ts]: Add helper

**Comment by samelhusseini** [comment]
Agreed, we need to do a deep test of this.

File: webapp/src/app.tsx (hunk 2 above)
Comment (samelhusseini) [reply webapp/src/app.tsx]: It relies on resourceImporters which is attached to the editor (for extensions), so we'd have to pass it around or just keep the method there.

File: webapp/src/codecard.tsx (hunk 3 above)
Comment (samelhusseini) [reply webapp/src/codecard.tsx]: Not sure, can you give an example?

Comment (samelhusseini) [reply webapp/src/codecard.tsx]: (also this is for codecard.tsx which is the react equivalent of the docrender one) and only used in the editor. I don't believe we support youtube embeds in the editor.

Comment (pelikhan) [reply webapp/src/codecard.tsx]: It’s used in docs but I don’t think we load the YouTube iframe in the editor.

File: webapp/src/simulator.ts (hunk 4 above)
Comment (samelhusseini) [reply webapp/src/simulator.ts]: Done.

File: package.json (hunk 1 above)
Comment (samelhusseini) [reply package.json]: Two options that we don't use had been removed. https://github.com/krisk/Fuse/blob/master/CHANGELOG.md

File: webapp/src/codecard.tsx (hunk 3 above)
Comment (samelhusseini) [reply webapp/src/codecard.tsx]: Cool. I think it's fine to get rid of it then.

**Comment by pelikhan** [comment]
Looks like debugger UI is in here too. Probably want to merge that one first.

**Comment by samelhusseini** [comment]
Where do you see the debugger UI. I made sure of not including it in?

**Comment by samelhusseini** [comment]
If you're referring to the changes in simulator.ts (debuggerDOM). Those are unrelated, and basically are the old debugger UI. (the one you access from ?dbg=1). I'll be removing these once things settle down.

**Comment by matthewtebbs** [comment]
FYI I created a local repo with this change merged into it. So far I do see an issue of the System menu on the top-right of the title bar being anchored in the wrong direction, e.g. it is clipped to the right side of the browser window. 

With this PR it happens in Firefox and Chrome (see below).

image

It does not happen with a version of pxt-core without this PR (see below).

image

**Comment by samelhusseini** [comment]
Oh thanks for doing that Matt, I’ll check on that issue tonight.

________________________________

**Comment by samelhusseini** [comment]
That's been fixed, let me know if you find any other issues @matthewtebbs

**Comment by matthewtebbs** [comment]
FWIW this PR is awesome. I've been running with it on a private fork and the user experience is definitely zippier.

**Comment by samelhusseini** [comment]
Thanks!

**Comment by abchatra** [comment]
Can you share the performance number?

**Comment by matthewtebbs** [comment]
Likewise. Happy to help test on Ubuntu, macOS and Windows 10. Firefox, Chrome, Edge and IE (as installed on Windows 10).

**Comment by samelhusseini** [comment]
Thanks heaps for the enthusiasm to help out with the testing process. I'm still figuring out a couple of known issues (see above list) before getting this in, but in the meantime, here's a link to a build off master with this change that you can use for testing:
https://makecode.adafruit.com/app/fde30de22b95b36daa110531946c11376ded0ff0-80cc42678c

**Comment by samelhusseini** [comment]
Any word on testing efforts here? Anyone find any issues?

**Comment by matthewtebbs** [comment]
Gave it a quick look on macOS on Safari, Chrome, Firefox latest versions. Found some minor bugs what may not be related to this change.

**Comment by samelhusseini** [comment]
CTI found these accessibility regressions: 
- [ ] Can't navigate to 'Privacy & Cookies', 'Terms Of Use' and 'About' buttons
- [ ] The focus returns to the first element of the Editor page after closing the function under Dropdown
- [ ] Entering “Exit Project” modal, first interactable element is not focused
- [ ] Entering “Projects Settings” page, no element is focused

**Comment by matthewtebbs** [comment]
Didn't test for accessibility. Next time I will. You get what you pay for :)

**Comment by samelhusseini** [comment]
Once I fix the accessibility issues, I think this will be ready to merge.

**Comment by matthewtebbs** [comment]
Me too.

________________________________

**Pull request merged**